pkg/cloudprovider/providers/vsphere/vclib/virtualdisk.go

```diff
@@ -0,0 +1,51 @@
+package vclib
+
+import (
+	"github.com/golang/glog"
+	"github.com/vmware/govmomi/object"
+	"golang.org/x/net/context"
+)
+
+// VirtualDisk is for the Disk Management
+type VirtualDisk struct {
+	DiskPath      string
+	VolumeOptions VolumeOptions
+	VMOptions     VMOptions
+}
+
+// Disk defines interfaces for creating disk
+type Disk interface {
+	Create(ctx context.Context, datastore *Datastore) error
+}
+
+// GetDiskManager returns vmDiskManager or vdmDiskManager based on given volumeoptions
+func GetDiskManager(disk *VirtualDisk) Disk {
```

Comment:
I think if you pass operation along with virtualDisk then for delete also you can make use of the factory.
IMHO below parameters in factory with dummy implementation will be extensible..


func GetDiskManager(disk *VirtualDisk, op VirtualDiskOperation) Disk {
    if op == "create" {
        // your current code
    } else if op == "delete" {
       // currently just return VDM but very soon VStorageObjectManager needs to be returned based on disk.diskId
   }
}